src/controller/python/test/mobile-device-test.py

```diff
@@ -0,0 +1,169 @@
+# Network id, for the thread network, current a const value, will be changed to XPANID of the thread network.
+TEST_THREAD_NETWORK_ID = "0123456789abcdef"
+
+
+def TestFail(message):
+    logger.fatal("Testfail: {}".format(message))
+    os._exit(1)
+
+
+def FailIfNot(cond, message):
+    if not cond:
+        TestFail(message)
+
+
+class TestTimeout(threading.Thread):
+    def __init__(self, timeout: int):
+        threading.Thread.__init__(self)
+        self._timeout = timeout
+        self._should_stop = False
+        self._cv = threading.Condition()
+
+    def stop(self):
+        with self._cv:
+            self._should_stop = True
+            self._cv.notify_all()
+        self.join()
+
+    def run(self):
+        stop_time = time.time() + self._timeout
+        logger.info("Test timeout set to {} seconds".format(self._timeout))
+        with self._cv:
+            wait_time = stop_time - time.time()
+            while wait_time > 0 and not self._should_stop:
+                self._cv.wait(wait_time)
+                wait_time = stop_time - time.time()
+        if time.time() > stop_time:
+            TestFail("Timeout")
+
+
+class MobileDeviceTests():
+    def __init__(self, nodeid: int):
+        self.devCtrl = ChipDeviceCtrl.ChipDeviceController(
+            controllerNodeId=nodeid)
+        self.logger = logger
+
+    def TestKeyExchange(self, ip: str, setuppin: int, nodeid: int):
+        self.logger.info("Conducting key exchange with device {}".format(ip))
+        if not self.devCtrl.ConnectIP(ip.encode("utf-8"), setuppin, nodeid):
+            self.logger.info(
+                "Failed to finish key exchange with device {}".format(ip))
+            return False
+        self.logger.info("Device finished key exchange.")
+        return True
+
+    def TestNetworkCommissioning(self, nodeid: int):
+        self.logger.info("Commissioning network to device {}".format(nodeid))
+        try:
+            self.devCtrl.ZCLSend("NetworkCommissioning", "AddThreadNetwork", nodeid, 1, 0, {
+                "operationalDataset": bytes.fromhex(TEST_THREAD_NETWORK_DATASET_TLV),
+                "breadcrumb": 0,
+                "timeoutMs": 1000})
+        except Exception as ex:
+            self.logger.exception("Failed to send AddThreadNetwork command")
+            return False
+        # TODO: ZCLSend should be synchronous, currently, IM does not pass IM status response to upper layer.
+        time.sleep(2)
+        self.logger.info("Send EnableNetwork command to device {}".format(nodeid))
+        try:
+            self.devCtrl.ZCLSend("NetworkCommissioning", "EnableNetwork", nodeid, 1, 0, {
+                "networkID": bytes.fromhex(TEST_THREAD_NETWORK_ID),
+                "breadcrumb": 0,
+                "timeoutMs": 1000})
+        except Exception as ex:
+            self.logger.exception("Failed to send EnableNetwork command")
+            return False
+        # Wait for thread network join
+        time.sleep(45)
```

Comment:
nit: What is the reason to wait 45 seconds for joining, it seems too much.

Reply:
Generally we should avoid all sleeps. Could we poll for some sort status change or away a 'joined' notification?

Reply:
That kind of implementation will be added in later IM & cluster updates. For now I think we can set it to 15 seconds, which is the time we used in SetupThreadNetwork in Cirque tests.

Reply:
Added a comment for which feature is required before we can waiting for network join in a CHIP way.